contracts/cw20-base/src/allowances.rs

```diff
@@ -0,0 +1,331 @@
+use cosmwasm_std::{
+    Addr, Binary, Deps, DepsMut, Env, MessageInfo, Order, Response, StdError, StdResult, Uint128,
+};
+use cw20_allowances::responses::{
+    AllAllowancesResponse, AllSpenderAllowancesResponse, AllowanceInfo, AllowanceResponse,
+    SpenderAllowanceInfo,
+};
+use cw20_allowances::Cw20Allowances;
+use cw_storage_plus::{Bound, Bounder};
+use cw_utils::Expiration;
+
+use crate::contract::Cw20Base;
+use crate::error::ContractError;
+use crate::responses::Cw20ReceiveMsg;
+
+// settings for pagination
+const MAX_LIMIT: u32 = 30;
+const DEFAULT_LIMIT: u32 = 10;
+
+impl Cw20Allowances for Cw20Base<'_> {
+    type Error = ContractError;
+
+    /// Allows spender to access an additional amount tokens from the owner's (env.sender) account.
+    /// If expires is Some(), overwrites current allowance expiration with this one.
+    fn increase_allowance(
+        &self,
+        ctx: (DepsMut, Env, MessageInfo),
+        spender: String,
+        amount: Uint128,
+        expires: Option<Expiration>,
+    ) -> Result<Response, Self::Error> {
+        let (deps, env, info) = ctx;
+
+        let spender_addr = deps.api.addr_validate(&spender)?;
+        if spender_addr == info.sender {
+            return Err(ContractError::CannotSetOwnAccount {});
+        }
+
+        let update_fn = |allow: Option<AllowanceResponse>| -> Result<_, _> {
+            let allow = allow.unwrap_or_default();
+            let allowance = allow.allowance + amount;
+            match expires {
+                Some(expires) if !expires.is_expired(&env.block) => {
+                    Ok(AllowanceResponse { allowance, expires })
+                }
+                None => Ok(AllowanceResponse { allowance, ..allow }),
+                _ => Err(ContractError::InvalidExpiration {}),
+            }
+        };
+        self.allowances
+            .update(deps.storage, (&info.sender, &spender_addr), update_fn)?;
+        self.allowances_spender
+            .update(deps.storage, (&spender_addr, &info.sender), update_fn)?;
+
+        let res = Response::new()
+            .add_attribute("action", "increase_allowance")
+            .add_attribute("owner", info.sender)
+            .add_attribute("spender", spender)
+            .add_attribute("amount", amount);
+        Ok(res)
+    }
+
+    /// Lowers the spender's access of tokens from the owner's (env.sender) account by amount.
+    /// If expires is Some(), overwrites current allowance expiration with this one.
+    fn decrease_allowance(
+        &self,
+        ctx: (DepsMut, Env, MessageInfo),
+        spender: String,
+        amount: Uint128,
+        expires: Option<Expiration>,
+    ) -> Result<Response, Self::Error> {
+        let (deps, env, info) = ctx;
+
+        let spender_addr = deps.api.addr_validate(&spender)?;
```

Comment:
Minor, but you don't have to validate this address. Validation takes time (and gas), but if the address is not valid, it would for sure fail on loading allowance (as there is no way an invalid address is inserted into the map). We have `Addr::unchecked(&spender)` for such cases.